File: implementations/rust/osf-core/examples/inspect.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Optimeas GmbH

//! `inspect` — minimal CLI that prints what `osf-core` knows about a file.
//!
//! Usage:
//!
//! ```text
//! cargo run --example inspect -- <path-to-osf-file>
//! ```
//!
//! Prints the magic header, the file-level metadata recorded in the
//! metablock, and a one-line summary per channel. Block-stream
//! inspection follows in later sessions.
//!
//! Run with `RUST_LOG=debug cargo run --example inspect -- <path>` to
//! see diagnostics about deprecated fields, accepted alternatives, and
//! unknown attributes.

use osf_core::{
    ChannelType, DataType, MetaBlock, MetaChannel, compression, parse_magic_header,
    parse_metablock,
};
use std::env;
use std::fs::File;
use std::io::Read;
use std::process::ExitCode;

fn main() -> ExitCode {
    env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("warn"))
        .format_timestamp(None)
        .init();

    let mut args = env::args().skip(1);
    let Some(path) = args.next() else {
        eprintln!("usage: inspect <path-to-osf-file>");
        return ExitCode::from(2);
    };

    let file = match File::open(&path) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("cannot open {path}: {e}");
            return ExitCode::FAILURE;
        }
    };

    // OSFZ-aware: detect gzip / zlib wrappers transparently. Plain
    // OSF files come through as MaybeCompressed::Plain and continue
    // unchanged.
    let mut stream = match compression::detect_and_wrap(file) {
        Ok(s) => s,
        Err(e) => {
            eprintln!("{path}: cannot detect compression: {e}");
            return ExitCode::FAILURE;
        }
    };
    let compression_format = stream.detected_format();

    let header = match parse_magic_header(&mut stream) {
        Ok(h) => h,
        Err(e) => {
            eprintln!("{path}: {e}");
            return ExitCode::FAILURE;
        }
    };

    let mut body = vec![0u8; header.metablock_len as usize];
    if let Err(e) = stream.read_exact(&mut body) {
        eprintln!("{path}: cannot read metablock: {e}");
        return ExitCode::FAILURE;
    }

    let metablock = match parse_metablock(header.version, &body) {
        Ok(mb) => mb,
        Err(e) => {
            eprintln!("{path}: metablock parse failed: {e}");
            return ExitCode::FAILURE;
        }
    };

    print_summary(&path, &header, &metablock, compression_format);
    ExitCode::SUCCESS
}

fn print_summary(
    path: &str,
    header: &osf_core::MagicHeader,
    mb: &MetaBlock,
    compression_format: compression::CompressionFormat,
) {
    use compression::CompressionFormat::{Gzip, None as Plain, Zlib};
    println!("path:           {path}");
    let compressed_label = match compression_format {
        Plain => "no",
        Zlib => "yes (zlib)",
        Gzip => "yes (gzip)",
    };
    println!("compressed:     {compressed_label}");
    println!("version:        {:?}", header.version);
    println!("metablock_len:  {} bytes", header.metablock_len);
    println!(
        "created_utc:    {}",
        mb.file_info.created_utc.as_deref().unwrap_or("-")
    );
    println!(
        "creator:        {}",
        mb.file_info.creator.as_deref().unwrap_or("-")
    );
    println!("channels:       {}", mb.channels.len());

    if !mb.channels.is_empty() {
        let max_name = mb
            .channels
            .iter()
            .map(|c| c.name.chars().count())
            .max()
            .unwrap_or(0)
            .min(48);
        for chan in &mb.channels {
            println!("   {}", format_channel(chan, max_name));
        }
    }

    println!("infos:          {}", mb.infos.len());
}

fn format_channel(chan: &MetaChannel, name_width: usize) -> String {
    let display_name = truncate_to(&chan.name, name_width);
    let ct = format_channel_type(&chan.channel_type);
    let dt = format_data_type(&chan.data_type);
    let unit = chan
        .physical_unit
        .as_deref()
        .filter(|u| !u.is_empty())
        .unwrap_or("-");
    let incr = chan
        .time_increment_ns
        .map(|n| format!("incr_ns={n}"))
        .unwrap_or_else(|| "incr_ns=-".to_string());
    format!(
        "[{:>3}] {:<width$}  {:<11}  {:<8}  unit={:<8}  {}",
        chan.index,
        display_name,
        ct,
        dt,
        unit,
        incr,
        width = name_width
    )
}

fn truncate_to(s: &str, max: usize) -> String {
    if s.chars().count() <= max {
        s.to_string()
    } else {
        let mut out: String = s.chars().take(max.saturating_sub(1)).collect();
        out.push('…');
        out
    }
}

fn format_channel_type(ct: &ChannelType) -> String {
    match ct {
        ChannelType::Scalar => "scalar".into(),
        ChannelType::Vector => "vector".into(),
        ChannelType::Matrix => "matrix".into(),
        ChannelType::Binary => "binary".into(),
        ChannelType::Unsupported(s) => format!("?{s}"),
    }
}

fn format_data_type(dt: &DataType) -> String {
    match dt {
        DataType::Bool => "bool".into(),
        DataType::Int8 => "int8".into(),
        DataType::Int16 => "int16".into(),
        DataType::Int32 => "int32".into(),
        DataType::Int64 => "int64".into(),
        DataType::UInt8 => "uint8".into(),
        DataType::UInt16 => "uint16".into(),
        DataType::UInt32 => "uint32".into(),
        DataType::UInt64 => "uint64".into(),
        DataType::Float => "float".into(),
        DataType::Double => "double".into(),
        DataType::String => "string".into(),
        DataType::Binary | DataType::ByteArray => "binary".into(),
        DataType::GpsLocation => "gpsloc".into(),
        DataType::Unsupported(s) => format!("?{s}"),
    }
}

File: implementations/cpp/examples/inspect.cpp
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Optimeas GmbH
//
// inspect — print a summary of an OSF / OSFZ file.
//
// Usage:
//   inspect <file>
//
// Prints file-level metadata (version, creator, createdUtc, compression)
// and a one-line summary per channel.

#include <osf/osf.h>

#include <algorithm>
#include <cstddef>
#include <iomanip>
#include <iostream>
#include <string>

namespace {

// ── Helpers ───────────────────────────────────────────────────────────────

std::string dataTypeName(osf::DataType dt) {
    switch (dt) {
        case osf::DataType::Bool:        return "bool";
        case osf::DataType::Int8:        return "int8";
        case osf::DataType::Int16:       return "int16";
        case osf::DataType::Int32:       return "int32";
        case osf::DataType::Int64:       return "int64";
        case osf::DataType::UInt8:       return "uint8";
        case osf::DataType::UInt16:      return "uint16";
        case osf::DataType::UInt32:      return "uint32";
        case osf::DataType::UInt64:      return "uint64";
        case osf::DataType::Float:       return "float";
        case osf::DataType::Double:      return "double";
        case osf::DataType::String:      return "string";
        case osf::DataType::Binary:      return "binary";
        case osf::DataType::ByteArray:   return "binary";
        case osf::DataType::GpsLocation: return "gpsloc";
        default:                         return "?unsupported";
    }
}

std::string channelTypeName(osf::ChannelType ct) {
    switch (ct) {
        case osf::ChannelType::Scalar: return "scalar";
        case osf::ChannelType::Vector: return "vector";
        case osf::ChannelType::Matrix: return "matrix";
        case osf::ChannelType::Binary: return "binary";
        default:                       return "?unsupported";
    }
}

// Truncate a UTF-8 string to at most maxChars characters (by byte for
// simplicity; only the channel-name column is affected).
std::string truncate(std::string const& s, std::size_t maxChars) {
    if (s.size() <= maxChars) return s;
    return s.substr(0, maxChars - 1) + "~";
}

} // namespace

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "usage: inspect <file>\n";
        return 2;
    }

    std::string const path = argv[1];

    auto result = osf::DataManager::loadFromFile(path);
    if (!result) {
        std::cerr << path << ": " << result.error().message << "\n";
        return 1;
    }

    osf::DataManager const& mgr = *result;
    osf::ReaderStats  const& st  = mgr.stats;
    osf::FileInfo     const& fi  = mgr.meta.fileInfo;

    // ── File-level summary ──────────────────────────────────────────────
    std::cout << "path:           " << path << "\n";

    // Compression
    std::string compressedLabel;
    if (!st.compressed) {
        compressedLabel = "no";
    } else {
        switch (st.compressionFormat) {
            case osf::CompressionFormat::Gzip: compressedLabel = "yes (gzip)"; break;
            case osf::CompressionFormat::Zlib: compressedLabel = "yes (zlib)"; break;
            default:                           compressedLabel = "yes";         break;
        }
    }
    std::cout << "compressed:     " << compressedLabel << "\n";
    std::cout << "version:        " << fi.version << "\n";
    std::cout << "creator:        " << fi.creator.value_or("-") << "\n";
    std::cout << "created_utc:    " << fi.createdUtc.value_or("-") << "\n";

    std::size_t const nchan = mgr.channels().size();
    std::cout << "channels:       " << nchan << "\n";

    // ── Per-channel lines ───────────────────────────────────────────────
    if (nchan > 0) {
        // Determine column width for the name field (capped at 48).
        std::size_t maxName = 0;
        for (osf::DataChannel const& dc : mgr.channels()) {
            maxName = std::max(maxName, osf::channelName(dc).size());
        }
        maxName = std::min(maxName, std::size_t{48});

        for (osf::DataChannel const& dc : mgr.channels()) {
            std::uint16_t const idx   = osf::channelIndex(dc);
            std::string   const name  = truncate(osf::channelName(dc), maxName);
            std::string   const ct    = channelTypeName(osf::channelMeta(dc).channelType);
            std::string   const dt    = dataTypeName(osf::channelDataType(dc));
            std::string   const unit  = osf::channelPhysicalUnit(dc).value_or("-");

            std::cout << "   ["
                      << std::setw(3) << std::right << idx
                      << "] "
                      << std::setw(static_cast<int>(maxName)) << std::left << name
                      << "  "
                      << std::setw(11) << std::left << ct
                      << "  "
                      << std::setw(8) << std::left << dt
                      << "  unit=" << unit
                      << "\n";
        }
    }

    std::cout << "infos:          " << mgr.meta.infos.size() << "\n";
    return 0;
}
